### PR TITLE
packaging: use snapd type and snapcraft 3.x

### PR DESCRIPTION
**Pull request by stolowski**
Use `type: snapd` in `snapcraft.yaml` and update snap packging to use `build-aux` directory supported by newer snapcraft. Also specific "base:" is now required. This requires snapcraft with "snapd" type support that is currently in **edge**.

Many thanks to @cmatsuoka for hints and for his patch that I cherry-picked here, it fixed some fundamental issues I had with this change. I tested the build succesfuly with snapcraft 3.6+git23.g665143c.

Marking blocked until "snapd" type support in snapcraft hits stable, and the remaining snapd-type related PRs in snapd land (tests will fail until then anyway).

### CODE REVIEW

**Comment by stolowski** [comment]
I got a word from Sergio, the next stable release of snapcraft is expected in the week after Jul 15th.

**Comment by stolowski** [comment]
> Has `snapcraft.yaml` also been placed in `build-aux/snap` along with the local plugin?

@cmatsuoka yes, and snapcraft looks it up there:

File: build-aux/snap/plugins/x_builddeb.py [hunk 1]
```diff
@@ -17,33 +17,15 @@
...
-
 class XBuildDeb(snapcraft.BasePlugin):
 
     def build(self):
         super().build()
-        self.run(["sudo", "apt-get", "build-dep", "-y", "./"])
         # ensure we have go in our PATH
         env=os.environ.copy()
+        env["DEBIAN_FRONTEND"] = "noninteractive"
+        env["DEBCONF_NONINTERACTIVE_SEEN"] = "true"
```
Comment (mvo5) [inline build-aux/snap/plugins/x_builddeb.py]: Nice, thank you!

**Comment by mvo5** [comment]
We need to tweak the snapd build test - it is using the snapcraft deb which is still at version 2.
```
...
Setting up snapcraft (2.43.1) ...

Processing triggers for libc-bin (2.23-0ubuntu11) ...

+ cd /home/gopath/src/github.com/snapcore/snapd
+ snapcraft
Could not find snap/snapcraft.yaml. Are you sure you are in the right directory?
To start a new project, use `snapcraft init`
```

**Comment by mvo5** [comment]
As an experiment I switched to:
```
index 8fca4b6dd8..de312b5160 100644
--- a/tests/main/snapd-snap/task.yaml
+++ b/tests/main/snapd-snap/task.yaml
@@ -11,16 +11,20 @@ restore: |
     echo "Cleanup build artifacts"
     not command -v snapcraft >/dev/null || snapcraft clean
     echo "Cleanup the installed snapcraft"
-    apt autoremove -y snapcraft
+    snap remove --purge lxd snapcraft
     echo "Cleanup the build snapd snap"
     rm -f snapd_*.snap
 
 execute: |
+    echo "Installing lxd"
+    snap install lxd
+    lxd init --auto
+
     echo "Installing snapscraft"
-    apt install -y snapcraft
+    snap install --classic --edge snapcraft
     # shellcheck disable=SC2164
     cd "$PROJECT_PATH"
-    snapcraft
+    snapcraft --use-lxd
 
     echo "Ensure we have the fc-cache binaries"
     unsquashfs -ll snapd_*.snap | MATCH bin/fc-cache-v6
```
but even with the snapcraft 3 edge the build fails very early:
```
qemu:ubuntu-16.04-64 ...# snap run snapcraft  --use-lxd
Using 'build-aux/snap/snapcraft.yaml': Project assets will be searched for from the 'build-aux/snap' directory.
The snap/ directory is meant specifically for snapcraft, but it contains the following non-snapcraft-related paths, which is unsupported and will cause unexpected behavior:
- broken.go
- broken_test.go
- channel.go
- channel_test.go
- container.go
- container_test.go
- epoch.go
- epoch_test.go
- errors.go
- export_test.go
- gadget.go
- gadget_test.go
- hooktypes.go
- hotplug_key.go
- hotplug_key_test.go
- implicit.go
- implicit_test.go
- info.go
- info_snap_yaml.go
- info_snap_yaml_test.go
- info_test.go
- naming
- naming/naming_test.go
- naming/validate.go
- naming/validate_test.go
- pack
- pack/export_test.go
- pack/pack.go
- pack/pack_test.go
- restartcond.go
- restartcond_test.go
- revision.go
- revision_test.go
- seed_yaml.go
- seed_yaml_test.go
- snapdir
- snapdir/snapdir.go
- snapdir/snapdir_test.go
- snapenv
- snapenv/snapenv.go
- snapenv/snapenv_test.go
- snaptest
- snaptest/snaptest.go
- snaptest/snaptest_test.go
- squashfs
- squashfs/export_test.go
- squashfs/squashfs.go
- squashfs/squashfs_test.go
- squashfs/stat.go
- squashfs/stat_test.go
- types.go
- types_test.go
- validate.go
- validate_test.go

If you must store these files within the snap/ directory, move them to snap/local/, which is ignored by snapcraft.
The LXD provider is offered as a technology preview for early adopters.
The command line interface, container names or lifecycle handling may change in upcoming releases.
Launching a container.
Waiting for cloud-init
.............
status: done
error: no changes of type "auto-refresh" found
error: cannot find signatures with metadata for snap "/var/tmp/core.snap"
Sorry, an error occurred in Snapcraft:
An error occurred when trying to execute 'snap install /var/tmp/core.snap' with 'LXD': returned exit code 1.
```
slightly sad, but it seems snapcraft 3 is not quite ready for us yet :/

**Comment by mvo5** [comment]
We can use the pass-through to set the snapd type even in snapcraft 2.

**Comment by stolowski** [comment]
Hmm, bummer, it was working when I tested it manually in a VM 🙁. The warning about snap/ is harmless AFICT; the error at the end looks weird and not directly related, is it? Maybe a transient error?

**Comment by jhenstridge** [comment]
> error: cannot find signatures with metadata for snap "/var/tmp/core.snap"

I suspect this error is due to Snapcraft trying to save bandwidth by copying the host system's core snap into the LXD container.  That's going to be the snap built by the spread tests, which isn't going to be signed by the store.

One possibility would be to use `--destructive-mode` rather than `--use-lxd`.  This behaves similar to Snapcraft 2.x and performs the build on the host system rather than a VM or container, which might be appropriate for the Spread test.

**Comment by stolowski** [comment]
Many thanks to @jhenstridge and @anonymouse64 for inisghts into the root cause of the failure and suggestions. `--destructive-mode` flag for snapcraft indeed made this particular spread test happy (in my manual tests at least) - let's see what travis thinks. Thank you!

**Comment by stolowski** [comment]
@mvo @pedronis This is ready for re-review. Note, I've reverted the `snap: remove magic to transform name "snapd" to TypeSnapd` commit as we have at least one spread test that downloads snapd snap from edge & re-packs it (instead of re-builing with snapcraft as we do in `snapd-snap` spread test), so we need to land new snapd snap in edge first and the "magic" can be removed afterwards in separate PR.

**Comment by jhenstridge** [comment]
Glad that suggestion helped.  On a semantic level, having `base: core` in the snapd snap seems wrong: it doesn't actually depend on a base snap, and on an Ubuntu Core 18 device, you won't even necessarily have the core snap installed.

This seems like a case where the new `build-base` keyword would make more sense.  Something like `build-base: core` would mean "pick the build environment associated with `core`, even though I haven't written `base: core`".  Unfortunately, that feature seems to be restricted to `type: base` snaps at the moment.  I'm hopeful that will change though.

**Comment by stolowski** [comment]
Marking blocked again, we need `build-base: core` with `type: snapd`, and this requires a change in spancraft.

File: build-aux/snap/snapcraft.yaml [hunk 2]
```diff
@@ -10,9 +10,9 @@ description: |
 version: set-by-version-script-thxbye
 version-script: |
   ./mkversion.sh --output-only
-#FIXME: enable once snapcraft understands this
-#type: snapd
+type: snapd
 grade: stable
+base: core
```
Comment (pedronis) [inline build-aux/snap/snapcraft.yaml]: ideally this should be a build-base instead as we know

File: tests/main/snapd-snap/task.yaml [hunk 3]
```diff
@@ -9,18 +9,18 @@ priority: 100
 restore: |
     cd "$PROJECT_PATH"
     echo "Cleanup build artifacts"
-    not command -v snapcraft >/dev/null || snapcraft clean
+    rm -rf parts
```
Comment (pedronis) [inline tests/main/snapd-snap/task.yaml]: not sure why this change, if there was no snapcraft there would be no parts either, no?

Comment (stolowski) [reply tests/main/snapd-snap/task.yaml]: This was because of issues with "clean" when `--destructive-mode` flag was used for the build, and before realizing `clean` also accepts `--destructive-mode` flag. Fixed.

File: build-aux/snap/snapcraft.yaml (hunk 2 above)
Comment (stolowski) [reply build-aux/snap/snapcraft.yaml]: Done now.

**Comment by anonymouse64** [comment]
Since we don't need to patch snapcraft anymore, could we eliminate the custom plugin and use `override-build` and the `make plugin` instead to implement the build? That seems like it would be more maintainable than having a custom plugin IMHO. Not necessarily for this PR however

**Comment by anonymouse64** [comment]
@stolowski I think this is unblocked from snapcraft upstream

**Comment by pedronis** [comment]
Need to block it again, there is no way to go through the transition ATM in the store.

**Comment by anonymouse64** [comment]
As per SU this morning and discussing over IRC, most of these changes have been implemented in https://github.com/snapcore/snapd/pull/7904, and when ready we can just submit a small PR enabling `type: snapd` whenever the store is ready for that.
